### PR TITLE
refactor(sdk)!: unify state transition processing

### PR DESCRIPTION
**Pull request by lklimek**
## Issue being fixed or feature implemented

Some code in state transition broadcast logic is very similar. This makes maintenance a bit more difficult, and things like retry logic harder to implement.

## What was done?

1. Added `Waitable` trait to wait for response and receive it from the server. Moved related code from direct implementations to `impl Waitable`.
2. Removed redundant code, using `BroadcastStateTransition` impl for `StateTransition` instead
3. Adjusted function arguments accordingly.
4. Adjusted default request settings in SdkBuilder to use 3 retries as default.

## How Has This Been Tested?

GHA

## Breaking Changes

* you need to `use Waitable` to call `wait_for_response()`
* changed multiple args of functions in state transition broadcast logic

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [ ] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

## Release Notes

- **New Features**
  - Introduced a new `Waitable` trait for improved response handling in state transitions.
  - Added optional `settings` parameters to various methods for enhanced configuration flexibility.

- **Bug Fixes**
  - Simplified error handling and response management across multiple traits and methods.

- **Refactor**
  - Streamlined the interfaces of `PurchaseDocument`, `PutContract`, `PutDocument`, `PutIdentity`, `TopUpIdentity`, `Transfer`, `TransferDocument`, and `UpdatePriceOfDocument` traits by removing redundant methods and consolidating logic.

- **Chores**
  - Added default request settings to the SDK for improved usability.

### CODE REVIEW

**Comment by lklimek** [comment]
@coderabbitai review all

File: packages/rs-sdk/src/platform/transition/purchase_document.rs [hunk 1]
```diff
@@ -1,23 +1,18 @@
...
 use dpp::state_transition::documents_batch_transition::methods::v0::DocumentsBatchTransitionMethodsV0;
 use dpp::state_transition::documents_batch_transition::DocumentsBatchTransition;
-use dpp::state_transition::proof_result::StateTransitionProofResult;
 use dpp::state_transition::StateTransition;
 
-use super::broadcast::BroadcastStateTransition;
-
 #[async_trait::async_trait]
 /// A trait for purchasing a document on Platform
 pub trait PurchaseDocument<S: Signer> {
```
Comment (shumkov) [inline packages/rs-sdk/src/platform/transition/purchase_document.rs:18]: Why you didn't add super trait Waitable here?

File: packages/rs-sdk/src/platform/transition/put_contract.rs [hunk 2]
```diff
@@ -10,14 +10,14 @@ use dpp::identity::signer::Signer;
...
-use dpp::state_transition::proof_result::StateTransitionProofResult;
 use dpp::state_transition::StateTransition;
 
 use super::broadcast::BroadcastStateTransition;
+use super::waitable::Waitable;
 
 #[async_trait::async_trait]
 /// A trait for putting a contract to platform
-pub trait PutContract<S: Signer> {
+pub trait PutContract<S: Signer>: Waitable {
```
Comment (shumkov) [inline packages/rs-sdk/src/platform/transition/put_contract.rs:20]: In other cases you don't add super trait so you need to use `<Self as Waitable>::`. I think this approach is cleaner

File: packages/rs-sdk/src/sdk.rs [hunk 3]
```diff
@@ -952,17 +962,18 @@ impl SdkBuilder {
     pub fn build(self) -> Result<Sdk, Error> {
         PlatformVersion::set_current(self.version);
 
+        let dapi_client_settings = self.settings.unwrap_or(DEFAULT_REQUEST_SETTINGS);
```
Comment (shumkov) [inline packages/rs-sdk/src/sdk.rs:965]: I think you need to merge passed setting on top of default setting, not replace. We can deal with settings in a separate PR. Let's discuss it and create a task.

File: packages/rs-sdk/src/platform/transition/waitable.rs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+use dpp::state_transition::StateTransition;
+use dpp::state_transition::StateTransitionLike;
+use dpp::ProtocolError;
+
+/// Waitable trait provides a wait to wait for a response of a state transition after it has been broadcast and
+/// receive altered objects.
+///
+/// This is simple conveniance trait wrapping the [`BroadcastStateTransition::wait_for_response`] method.
+#[async_trait::async_trait]
+pub trait Waitable: Sized {
```
Comment (shumkov) [inline packages/rs-sdk/src/platform/transition/waitable.rs:20]: I'm not sure this structure is the best, when we have all our entities used in one module. I think we should split code by domain entities: document related modules where we have everything about documents and so on. Let's discuss and create a task

**Review by shumkov — CHANGES_REQUESTED** [review]

**Review by shumkov — APPROVED** [review]

**Review by shumkov — APPROVED** [review]

**Pull request merged**